### PR TITLE
setoid use strict equality test optimisation (fixes #699)

### LINKED ISSUE #699: Speed up Setoid by optimistically test for strict Equality
When working on large data structure (like reducer states), this would help a lot.
@gcanti I may do a PR if you confirm interest beforehand.

### PR DESCRIPTION
**Pull request by sledorze**
**Before submitting a pull request,** please make sure the following is done:

- Fork [the repository](https://github.com/gcanti/fp-ts) and create your branch from `master`.
- Run `npm install` in the repository root.
- If you've fixed a bug or added code that should be tested, add tests!
- Ensure the test suite passes (`npm test`).

**Note**. If you find a typo in the **documentation**, make sure to modify the corresponding source (docs are generated).

### CODE REVIEW

**Comment by gcanti** [comment]
@sledorze this is error prone. Since it's always safe to check for strict equality while defining a `Setoid`, I propose to add a `fromEquals` helper (like `fromCompare` in `Ord`) to the `Setoid` module

```ts
// Setoid.ts

export const fromEquals = <A>(equals: (x: A, y: A) => boolean): Setoid<A> => {
  return {
    // optimization --v
    equals: (x, y) => x === y || equals(x, y)
  }
}
```

and then use it all over the place

```ts
// Identity.ts

export const getSetoid = <A>(setoid: Setoid<A>): Setoid<Identity<A>> => {
  return fromEquals((x, y) => setoid.equals(x.value, y.value))
}
```

**Comment by sledorze** [comment]
Honestly, I like the combinator but I fear the runtime overhead.
By error prone, what do you mean?

**Comment by gcanti** [comment]
> the runtime overhead

The inner call?

```ts
// this ---v
x === y || equals(x, y)
```

Seems negligible.

> By error prone, what do you mean?

"fix issue with wrong getProductSetoid optimisation"

**Comment by sledorze** [comment]
@gcanti I'll do the changes

**Comment by sledorze** [comment]
Way cleaner!

File: src/Const.ts [hunk 1]
```diff
@@ -47,11 +47,7 @@ export class Const<L, A> {
 /**
  * @since 1.0.0
  */
-export const getSetoid = <L, A>(S: Setoid<L>): Setoid<Const<L, A>> => {
-  return {
-    equals: (x, y) => S.equals(x.value, y.value)
-  }
-}
+export const getSetoid = <L, A>(S: Setoid<L>): Setoid<Const<L, A>> => fromEquals((x, y) => S.equals(x.value, y.value))
```
Comment (gcanti) [inline src/Const.ts]: @sledorze I'm sorry for the inconvenience but could you please keep the `{ return ... }` form for all functions? The docs generator must improved (it's on my radar) but the following short form is not handled at the moment 

Bad

```ts
export const getSetoid = <L, A>(S: Setoid<L>): Setoid<Const<L, A>> => fromEquals((x, y) => S.equals(x.value, y.value))
```

Good

```ts
export const getSetoid = <L, A>(S: Setoid<L>): Setoid<Const<L, A>> => {
  return fromEquals((x, y) => S.equals(x.value, y.value))
}
```

File: src/Either.ts [hunk 2]
```diff
@@ -277,13 +277,10 @@ export class Right<L, A> {
...
  * @since 1.0.0
  */
-export const getSetoid = <L, A>(SL: Setoid<L>, SA: Setoid<A>): Setoid<Either<L, A>> => {
-  return {
-    equals: (x, y) =>
-      x.isLeft() ? y.isLeft() && SL.equals(x.value, y.value) : y.isRight() && SA.equals(x.value, y.value)
-  }
-}
-
+export const getSetoid = <L, A>(SL: Setoid<L>, SA: Setoid<A>): Setoid<Either<L, A>> =>
```
Comment (gcanti) [inline src/Either.ts]: idem, etc.. etc..

Comment (sledorze) [reply src/Either.ts]: done

**Comment by sledorze** [comment]
@gcanti should be good now

File: src/Setoid.ts [hunk 3]
```diff
@@ -43,49 +50,35 @@ export const setoidBoolean: Setoid<boolean> = setoidStrict
 /**
  * @since 1.0.0
  */
-export const getArraySetoid = <A>(S: Setoid<A>): Setoid<Array<A>> => {
-  return {
-    equals: (xs, ys) => xs.length === ys.length && xs.every((x, i) => S.equals(x, ys[i]))
-  }
-}
+export const getArraySetoid = <A>(S: Setoid<A>): Setoid<Array<A>> =>
```
Comment (gcanti) [inline src/Setoid.ts]: missed :)

File: src/Setoid.ts [hunk 4]
```diff
@@ -43,49 +50,35 @@ export const setoidBoolean: Setoid<boolean> = setoidStrict
...
 export const getRecordSetoid = <O extends { [key: string]: any }>(
   setoids: { [K in keyof O]: Setoid<O[K]> }
-): Setoid<O> => {
-  return {
-    equals: (x, y) => {
-      for (const k in setoids) {
-        if (!setoids[k].equals(x[k], y[k])) {
-          return false
-        }
+): Setoid<O> =>
```
Comment (gcanti) [inline src/Setoid.ts]: idem

File: src/Setoid.ts [hunk 5]
```diff
@@ -43,49 +50,35 @@ export const setoidBoolean: Setoid<boolean> = setoidStrict
...
 /**
  * @since 1.0.0
  */
-export const getProductSetoid = <A, B>(SA: Setoid<A>, SB: Setoid<B>): Setoid<[A, B]> => {
-  return {
-    equals: ([xa, xb], [ya, yb]) => SA.equals(xa, ya) && SB.equals(xb, yb)
-  }
-}
-
+export const getProductSetoid = <A, B>(SA: Setoid<A>, SB: Setoid<B>): Setoid<[A, B]> =>
```
Comment (gcanti) [inline src/Setoid.ts]: idem

File: src/Setoid.ts [hunk 6]
```diff
@@ -43,49 +50,35 @@ export const setoidBoolean: Setoid<boolean> = setoidStrict
...
  * Returns the `Setoid` corresponding to the partitions of `B` induced by `f`
  *
  * @since 1.2.0
  */
-export const contramap = <A, B>(f: (b: B) => A, fa: Setoid<A>): Setoid<B> => {
-  return {
-    equals: on(fa.equals)(f)
-  }
-}
+export const contramap = <A, B>(f: (b: B) => A, fa: Setoid<A>): Setoid<B> => fromEquals((x, y) => fa.equals(f(x), f(y)))
```
Comment (gcanti) [inline src/Setoid.ts]: idem

**Comment by sledorze** [comment]
@gcanti may I push the button? :)

**Pull request merged**